contracts/sol6/fprConversionRates/SimpleVolumeImbalanceRecorder.sol

```diff
@@ -0,0 +1,150 @@
+        uint64 lastBlock;
+        int64 totalBuyUnitsImbalance;
+        uint64 lastRateUpdateBlock;
+    }
+
+    mapping(IERC20Ext => TokenControlInfo) internal tokenControlInfo;
+    mapping(IERC20Ext => TokenImbalanceData) public tokenImbalanceData;
+
+    constructor(address _admin) public Withdrawable(_admin) {}
+
+    function setTokenControlInfo(
+        IERC20Ext token,
+        uint128 minimalRecordResolution,
+        uint256 maxPerBlockImbalance,
+        uint256 maxTotalImbalance
+    ) external onlyAdmin {
+        require(minimalRecordResolution != 0, "zero minimalRecordResolution");
+        require(maxPerBlockImbalance != 0, "zero maxPerBlockImbalance");
+        require(maxTotalImbalance != 0, "zero maxTotalImbalance");
+
+        uint256 maxPerBlockImbalanceInResolution = maxPerBlockImbalance /
+            uint256(minimalRecordResolution);
+        // because abs(lastBlockBuyUnitsImbalance) <= MAX_INT64
+        // maxPerBlockImbalanceInResolution <= MAX_INT64
+        require(
+            maxPerBlockImbalanceInResolution <= uint256(MAX_INT64),
+            "overflow maxPerBlockImbalance"
+        );
+        uint256 maxTotalImbalanceInResolution = maxTotalImbalance /
+            uint256(minimalRecordResolution);
+        // because abs(totalBuyUnitsImbalance) <= MAX_INT64
+        // maxTotalImbalanceInResolution <= MAX_INT64
+        require(maxTotalImbalanceInResolution <= uint256(MAX_INT64), "overflow maxTotalImbalance");
+        tokenControlInfo[token] = TokenControlInfo(
+            uint128(minimalRecordResolution),
+            uint64(maxPerBlockImbalanceInResolution),
+            uint64(maxTotalImbalanceInResolution)
+        );
+    }
+
+    function getTokenControlInfo(IERC20Ext token)
+        external
+        view
+        returns (
+            uint256 minimalRecordResolution,
+            uint256 maxPerBlockImbalanceInResolution,
+            uint256 maxTotalImbalanceInResolution
+        )
+    {
+        minimalRecordResolution = tokenControlInfo[token].minimalRecordResolution;
+        maxPerBlockImbalanceInResolution = tokenControlInfo[token]
+            .maxPerBlockImbalanceInResolution;
+        maxTotalImbalanceInResolution = tokenControlInfo[token].maxTotalImbalanceInResolution;
+    }
+
+    function addImbalance(
+        IERC20Ext token,
+        int256 buyAmount,
+        uint256 rateUpdateBlock,
+        uint256 currentBlock
+    ) internal {
+        int64 recordedBuyAmount = safeInt64(
+            buyAmount / int256(tokenControlInfo[token].minimalRecordResolution)
+        );
+        TokenImbalanceData memory currentBlockData = tokenImbalanceData[token];
+
+        // first scenario - this is not the first tx in the current block
+        if (currentBlockData.lastBlock == currentBlock) {
+            currentBlockData.lastBlockBuyUnitsImbalance = currentBlockData
+                .lastBlockBuyUnitsImbalance
+                .add(recordedBuyAmount);
```

Comment:
again regarding using int64
need some analysis. a good chance we don't want a trade to fail due to imbalance recording.

Reply:
This limitation is similar to our sol4 version. 
Because both versions try to compress imbalanceData into 1 word, so this should be int64.

Do you want to change the limitation for imbalanceData in our new version